tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py

```diff
@@ -0,0 +1,166 @@
+    `minimize()` and `apply_gradients()`. 
+
+    Contrast to AdaMaxamOptimizer, the sparse implementation of this algorithm
+    (used when the gradient is an IndexedSlices object, typically because of
+    `tf.gather` or an embedding lookup in the forward pass) only updates
+    variable slices and corresponding `shadow_t` term when that part of
+    the variable was used in the forward pass. This means that the sparse
+    behavior is contrast to the dense behavior (similar to some momentum
+    implementations which ignore momentum unless a variable slice was actually
+    used).
+
+    Args:
+      learning_rate: A Tensor or a floating point value. The learning rate.
+      variance_parameter: A Tensor or a floating point value.
+        The variance control parameter.
+      num_workers: A int value. The number of workers.
+      use_locking: If True use locks for update operations.
+      name: Optional name for the operations created when applying gradients.
+        Defaults to "DelayCompensatedGradientDescentOptimizer".
+    """
+    num_workers = self._call_if_callable(num_workers)
+    if num_workers <= 0:
+      raise ValueError("num_workers must be positive: %s" % num_workers)
+    super(DelayCompensatedGradientDescentOptimizer, self).__init__(use_locking, name)
+    self._lr = learning_rate	
+    self._lambda = variance_parameter
+    self._num_workers = num_workers
+    self._learning_rate_tensor = None
+    self._lambda_tensor = None
+    self._use_locking = use_locking
+
+  def _create_slots(self, var_list):
+    for index in range(self._num_workers):
+      for v in var_list:
+        self._zeros_slot(v, "shadow_{0}".format(index), self._name)
+       
+       
+  def _prepare(self):
+    lr = self._call_if_callable(self._lr)
+    lambda_ = self._call_if_callable(self._lambda)
+
+    self._learning_rate_tensor = ops.convert_to_tensor(lr, name="learning_rate")
+    self._lambda_tensor = ops.convert_to_tensor(lambda_, name="lambda")
+
+  def _apply_dense(self, grad, var):
+  
+    shadow = self.get_slot(var, "shadow_{0}".format(self.worker_index))
+    return training_ops.apply_delay_compensated_gradient_descent(
+        var,
+        math_ops.cast(self._learning_rate_tensor, grad.dtype.base_dtype),
+        grad,
+        math_ops.cast(self._lambda_tensor, grad.dtype.base_dtype),
+        shadow,
+        use_locking=self._use_locking).op
+
+  def _resource_apply_dense(self, grad, var):
+   
+    shadow = self.get_slot(var, "shadow_{0}".format(self.worker_index))
+    return training_ops.resource_apply_delay_compensated_gradient_descent(
+        var.handle,
+        math_ops.cast(self._learning_rate_tensor, grad.dtype.base_dtype),
+        grad,
+        math_ops.cast(self._lambda_tensor, grad.dtype.base_dtype),
+        shadow.handle,
+        use_locking=self._use_locking)
+
+  def _apply_sparse_shared(self, grad, var, indices, scatter_add):
+   
+    shadow = self.get_slot(var, "shadow_{0}".format(self.worker_index))
```

Comment:
Should the shadow variables be checkpointed? Should they be stored in a parameter server or in the local worker?

Reply:
Shadow variables should be checkpointed because the corresponding parameters are checkpointed and the checkpointed shadow variables are more reasonable than the initialization values. Shadow variables are stored in the parameter server because workers have no idea about the updating parameters in the parameter server.